src/core/DataFile.cpp

```diff
@@ -1809,6 +1810,27 @@ void DataFile::upgrade_sampleAndHold()
 	}
 }
 
+//! Update MIDI CC indexes, so that they are counted from 0. Older releases of LMMS
+//! count the CCs from 1.
+void DataFile::upgrade_midiCCIndexing()
+{
+	std::vector<const char*> attributesToUpdate = {"inputcontroller", "outputcontroller"};
+
+	QDomNodeList elements = elementsByTagName("Midicontroller");
+	for(int i = 0; i < elements.length(); i++)
+	{
+		if (elements.item(i).isNull()) { continue; }
+		auto element = elements.item(i).toElement();
+		for (const char* attrName : attributesToUpdate)
+		{
+			if(element.hasAttribute(attrName))
+			{
+				int cc = std::stoi(element.attribute(attrName).toStdString());
```

Comment:
`QString` has a `toInt` method, which avoids the need to go via a `std::string`. (Then that header can go too.) It also defaults to zero on failure, which happens to be a sensible value here, since the attribute will be set to `NONE` in that case. On the other hand, `std::stoi` ignores an invalid suffix, and will throw an exception if there is no valid numeric prefix, which would need to be handled.

Reply:
done